examples/api-routes-apollo-server-and-client/lib/with-apollo-client.js

```diff
@@ -0,0 +1,108 @@
+import Head from 'next/head'
+import PropTypes from 'prop-types'
+import React from 'react'
+import { getDataFromTree, ApolloProvider } from 'react-apollo'
+import initApollo from './init-apollo'
+
+// Gets the display name of a JSX component for dev tools
+function getDisplayName (Component) {
+  return Component.displayName || Component.name || 'Unknown'
+}
+
+export default function withApollo (client, options = {}) {
+  if (!options.getDataFromTree) {
+    options.getDataFromTree = 'always'
+  }
+
+  return Page => {
+    return class WithApollo extends React.Component {
+      static displayName = `WithApollo(${getDisplayName(Page)})`
+
+      static propTypes = {
+        apolloState: PropTypes.object,
+        apollo: PropTypes.object
+      }
+
+      static getInitialProps = async ctx => {
+        const headers = ctx.req ? ctx.req.headers : {}
+        const apollo = initApollo(client, {
+          ctx,
+          headers,
+          ssrMode: typeof window === 'undefined'
+        })
+        const apolloState = {}
+        const getInitialProps = Page.getInitialProps
+
+        let pageProps = { pageProps: {} }
+
+        if (getInitialProps) {
+          ctx.apolloClient = apollo
+          pageProps = await getInitialProps(ctx)
+        }
+
+        if (ctx.res && (ctx.res.headersSent || ctx.res.finished)) {
+          return {}
+        }
+
+        if (
+          options.getDataFromTree === 'always' ||
+          (options.getDataFromTree === 'ssr' && typeof window === 'undefined')
+        ) {
+          try {
+            await getDataFromTree(
+              <ApolloProvider client={apollo}>
+                <Page
```

Comment:
Hi, we should probably update this to use `ctx.AppTree` once #8223 has landed. This will make sure `useRouter` still works here